loki/transformations/temporaries/stack_allocator.py

```diff
@@ -0,0 +1,1372 @@
+                arg_dims = (RangeIndex((None, None)),)
+                stack_arg_dict.setdefault(dtype, {})
+                stack_arg_dict[dtype][kind] = (stack_size_var, stack_var.clone(dimensions=arg_dims), stack_used_var)
+
+                # create stack_vars pair
+                stack_args += [stack_size_var,
+                               stack_var.clone(dimensions=stack_type.shape, type=stack_var.type.clone(contiguous=True)),
+                               stack_used_arg]
+                stack_vars += [stack_used_var]
+                pragma_string += f'{stack_var.name}, '
+
+        if pragma_string:
+            # remove ', '
+            pragma_string = pragma_string[:-2].lower()
+
+            present_pragma = None
+            acc_pragmas = [p for p in FindNodes(Pragma).visit(routine.body) if p.keyword.lower() == 'loki'] # acc
+            for pragma in acc_pragmas:
+                if pragma.content.lower().startswith('device-present'):
+                    present_pragma = pragma
+                    break
+            if present_pragma:
+                pragma_map = {present_pragma: None}
+                routine.body = Transformer(pragma_map).visit(routine.body)
+                content = re.sub(r'\bvars\(', f'vars({pragma_string}, ', present_pragma.content.lower())
+                present_pragma = present_pragma.clone(content = content)
+                pragma_data_end = None
+            else:
+                present_pragma = Pragma(keyword='loki', content=f'device-present vars({pragma_string})')
+                pragma_data_end = Pragma(keyword='loki', content='end device-present')
+
+            routine.body.prepend(present_pragma)
+            routine.body.append(pragma_data_end)
+        routine.body.prepend(as_tuple(assignments))
+        routine.variables += as_tuple(stack_vars)
+
+        # keep optional arguments last; a workaround for the fact that keyword arguments are not supported
+        # in device code
+        arg_pos = [routine.arguments.index(arg) for arg in routine.arguments if arg.type.optional]
+        if arg_pos:
+            routine.arguments = routine.arguments[:arg_pos[0]] + as_tuple(stack_args) + routine.arguments[arg_pos[0]:]
+        else:
+            routine.arguments += as_tuple(stack_args)
+
+        self.insert_stack_in_calls(routine, stack_arg_dict, successors)
+
+
+    def apply_pool_allocator_to_temporaries(self, routine, item=None): # pylint: disable=unused-argument
+        """
+        Apply raw stack allocator to local temporary arrays
+
+        This appends the relevant argument to the routine's dummy argument list and
+        creates the assignment for the local copy of the stack type.
+        For all local arrays, a Cray pointer is instantiated and the temporaries
+        are mapped via Cray pointers to the pool-allocated memory region.
+
+        The cumulative size of all temporary arrays is determined and returned.
```

Comment:
This docstring needs updating to reflect the base-class nature of the method here.